src/machines/java/com/enderio/machines/client/gui/widget/FluidTransformWidget.java

```diff
@@ -0,0 +1,70 @@
+import net.minecraft.client.gui.narration.NarrationElementOutput;
+import net.minecraft.client.renderer.texture.AbstractTexture;
+import net.minecraft.client.renderer.texture.TextureAtlas;
+import net.minecraft.client.renderer.texture.TextureAtlasSprite;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.util.FastColor;
+import net.minecraft.world.level.material.Fluid;
+import net.neoforged.neoforge.client.extensions.common.IClientFluidTypeExtensions;
+
+import java.util.function.Supplier;
+
+public class FluidTransformWidget extends EIOWidget {
+    private final Supplier<Boolean> isVisible;
+    private final Supplier<Fluid> first;
+    private final Supplier<Fluid> second;
+    private final Supplier<Float> progress;
+
+    public FluidTransformWidget(Supplier<Boolean> isVisible, Supplier<Fluid> first, Supplier<Fluid> second, Supplier<Float> progress, int x, int y, int width,
+        int height) {
+        super(x, y, width, height);
+        this.isVisible = isVisible;
+        this.first = first;
+        this.second = second;
+        this.progress = progress;
+    }
+
+    @Override
+    protected void renderWidget(GuiGraphics guiGraphics, int pMouseX, int pMouseY, float pPartialTick) {
+        RenderSystem.defaultBlendFunc();
+        RenderSystem.enableDepthTest();
+
+        if (isVisible.get()) {
+            renderFluid(guiGraphics, first.get(), 1 - progress.get());
+            renderFluid(guiGraphics, second.get(), progress.get());
+        }
+
+        RenderSystem.disableDepthTest();
+    }
+
+    @Override
+    protected void updateWidgetNarration(NarrationElementOutput pNarrationElementOutput) {}
+
+    public void renderFluid(GuiGraphics guiGraphics, Fluid fluid, float progress) {
+        Minecraft minecraft = Minecraft.getInstance();
+        IClientFluidTypeExtensions props = IClientFluidTypeExtensions.of(fluid);
+        ResourceLocation loc = props.getStillTexture();
+
+        AbstractTexture texture = minecraft.getTextureManager().getTexture(TextureAtlas.LOCATION_BLOCKS);
+        if (texture instanceof TextureAtlas atlas) {
+            TextureAtlasSprite sprite = atlas.getSprite(loc);
+
+            int color = props.getTintColor();
+            RenderSystem.setShaderColor(FastColor.ARGB32.red(color) / 255.0F, FastColor.ARGB32.green(color) / 255.0F, FastColor.ARGB32.blue(color) / 255.0F,
+                FastColor.ARGB32.alpha(color) * progress / 255.0F);
+            RenderSystem.enableBlend();
+
+            int atlasWidth = (int) (sprite.contents().width() / (sprite.getU1() - sprite.getU0()));
+            int atlasHeight = (int) (sprite.contents().height() / (sprite.getV1() - sprite.getV0()));
+            guiGraphics.blit(TextureAtlas.LOCATION_BLOCKS, x, y, width, height, sprite.getU0() * atlasWidth, sprite.getV0() * atlasHeight,
+                sprite.contents().width(), sprite.contents().height(), atlasWidth, atlasHeight);
+            RenderSystem.setShaderColor(1, 1, 1, 1);
```

Comment:
I think adding a disable blend after rendering is a good idea?

Reply:
Isn't blend one of the methods that is enabled by default? The default blend function calls just ensures that it is active and setup properly. I am not sure though as I don't have much rendering experience.